react/AddressForm.tsx

```diff
@@ -1,46 +1,156 @@
-import { LineFragment } from './typings/countryRulesTypes'
+import {
+  LineFragment,
+  Fields,
+  Field,
+  Display,
+} from './typings/countryRulesTypes.d'
+import rules from './countries/rules'
+import PlaceDetails from './PlaceDetails'
+import { FormattedMessage, defineMessages } from 'react-intl'
+import NumberOption from './components/NumberOption'
 
-interface Props {
-  address: Address
-  summary: LineFragment[][]
+const messages = defineMessages({
+  country: {
+    defaultMessage: '',
+    id: 'place-components.label.country',
+  },
+  state: {
+    defaultMessage: '',
+    id: 'place-components.label.state',
+  },
+  city: {
+    defaultMessage: '',
+    id: 'place-components.label.city',
+  },
+  neighborhood: {
+    defaultMessage: '',
+    id: 'place-components.label.neighborhood',
+  },
+  streetRoadAvenue: {
+    defaultMessage: '',
+    id: 'place-components.label.streetRoadAvenue',
+  },
+  numberOption: {
+    defaultMessage: '',
+    id: 'place-components.label.numberOption',
+  },
+  complement: {
+    defaultMessage: '',
+    id: 'place-components.label.complement',
+  },
+  stateAbbreviation: {
+    defaultMessage: '',
+    id: 'place-components.label.stateAbbreviation',
+  },
+  fieldRequired: {
+    defaultMessage: '',
+    id: 'place-components.error.fieldRequired',
+  },
+  edit: {
+    defaultMessage: '',
+    id: 'place-components.label.edit',
+  },
+})
+
+const getSummaryFields = (summary: LineFragment[][]) => {
+  let summaryFields = new Set()
+  summary.forEach((line: LineFragment[]) => {
+    line.forEach((fragment: LineFragment) => {
+      summaryFields.add(fragment.name)
+    })
+  })
+  return summaryFields
 }
 
-const AddressForm: StorefrontFunctionComponent<Props> = ({
-  address,
-  summary,
-}) => {
-  const { setAddress } = useAddressContext()
+const AddressForm: StorefrontFunctionComponent<{}> = () => {
+  const { address, setAddress } = useAddressContext()
+  const [editing, setEditing] = useState(false)
+  const { fields, display } = rules[address.country]
+  const summary = display.extended as LineFragment[][]
+
+  const [ignoredFields, setIgnoredFields] = useState(
+    getSummaryFields(display.compact)
+  )
+
+  const hasWithoutNumberOption = (label: string) => {
+    return label.length >= 6 && label.substr(label.length - 6) == 'Option'
+  }
+
+  const getFieldProps = (field: Field, fragment: LineFragment) => {
+    const { maxLength, autoComplete, required, label, options } = field
+
+    const onChange = (
+      event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
+    ) => {
+      setAddress({
+        ...address,
+        [fragment.name]: event.target.value,
+      })
+    }
+    const fieldRequired = {
+      errorMessage: <FormattedMessage {...messages.fieldRequired} />,
+    }
+    const value = address[fragment.name]
+
+    return {
+      label: (
+        <FormattedMessage {...messages[label as keyof (typeof messages)]} />
```

Comment:
@lucasecdb using `typeof messages` in the console only returns `object`. Would it be better to directly use `keyof object` in that case? or is there a way to define the type of that label more specifically?

Reply:
this is for typescript only, so it will correctly infer the type from the `messages` variable. in javascript land it will still be a regular object, but typescript will be able to tell the specific type

think doing `typeof SomeReactComponent`. in javascript, it will output `object`, but TS will be able to see it as a React component, that could include methods such as `componentDidMount`, etc

Reply:
you can create a [type alias](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases) so it will store the actual type of the variable.

```ts
type LabelType = keyof (typeof messages)
```

Reply:
Oh, great then, I removed the type alias since it was only used once, but if you think it's cleaner I can add it again, thank you!

Reply:
ok! no problem